### PR TITLE
fix(openclaw-plugin): notification, reconcile, config, and validation bugs

### PR DESCRIPTION
**Pull request by ggonzalez94**
## Summary

Audit-driven fixes for several bugs in the OpenClaw TAP plugin found by running a four-agent parallel code audit, then verifying each finding against the actual code before changing anything. Several of the audit's original findings turned out to be false positives on closer inspection — those are listed at the bottom for context.

## Fixes (each verified in code before editing)

- **Missing scheduling escalation notification (critical)** — `approveScheduling` returned `null` without pushing a notification, and the event classifier returns `null` for queued `action/request` (the hook is expected to own the notification lifecycle for queued requests, same as `approveTransfer`). Net effect: **deferred scheduling proposals were invisible to the operator**. Fixed in \`packages/openclaw-plugin/src/registry.ts\` by pushing an escalation notification + calling \`triggerEscalation\`, mirroring the transfer path.

- **Reconcile pileup** — the periodic reconcile \`setInterval\` had no in-flight guard. A slow \`syncOnce\` causes ticks to queue on the runtime mutex instead of being dropped. Added a \`reconcileInFlight\` flag that skips (with a warn log) when the previous run is still running.

- **OWS env overrides ignored in plugin** — CLI honors \`TAP_OWS_WALLET\`/\`TAP_OWS_API_KEY\` via \`config-loader.ts:53-54\`, but the plugin loaded config without passing them, so env-based wallet overrides were silently ignored inside Gateway. Now plumbed through both \`loadTrustedAgentConfigFromDataDir\` and \`createTapRuntime.configOptions\`.

- **Unbounded escalation notification queue** — \`TapNotificationQueue.evictIfNeeded\` allowed unbounded overflow once only escalations remained (a literal \`// Only escalations remain — allow overflow\` comment). If the operator never drained, memory grew without bound. Added a hard cap (\`maxSize * 2\`) with an \`onHardCapDrop\` hook; the registry logs each dropped escalation so it doesn't disappear silently. Constructor remains backwards-compatible with the bare-number form.

- **Amount validation for \`tap_gateway transfer\` / \`request_funds\`** — the plugin passed the raw string through to \`executeOnchainTransfer\`, so bad inputs (\`abc\`, \`-1\`, \`NaN\`, etc.) failed deep in viem with vague errors. Added \`requireAmount\` that rejects non-finite, non-positive, or non-numeric values early with a clear message.

- **Swallowed lock release errors** — \`ownerLock.release()\` failures in \`start()\`, \`stop()\`, and \`withTransportSession\` were silently ignored with \`.catch(() => {})\`. Replaced with \`this.log(\"warn\", ...)\` so stale-lock issues surface in Gateway logs.

- **\`transfer\` missing from skill docs** — \`tap_gateway transfer\` existed in code and inline docs but was missing from the action reference table in \`skills/trusted-agents/SKILL.md\`. Added the row (will propagate to the plugin skill mirror at build time).

## Tests

- \`packages/openclaw-plugin/test/notification-queue.test.ts\` — new tests for the hard cap behavior (drops oldest escalation once \`maxSize * 2\` is exceeded, fires the \`onHardCapDrop\` callback) and the legacy \`new TapNotificationQueue(3)\` constructor form.
- \`packages/openclaw-plugin/test/tool.test.ts\` — new tests covering \`requireAmount\`: rejects \`\"abc\"\`, rejects \`\"-1\"\`, and accepts well-formed decimals like \`\"0.5\"\`.

\`bun run typecheck && bun run lint && bun run test\` all green (683 tests passed, 38 skipped).

## Deliberately not fixed (audit false positives, verified and skipped)

These were flagged by the initial audit but turned out **not** to be bugs after verification:

- **\"Outbox poller not installed during \`withTransportSession\` reentry\"** — \`withTransportSession\` is only used for one-shot operations (\`syncOnce\`, \`processOutboxOnce\`) where \`runMaintenanceCycle\` processes the outbox inline. The background poller is intentionally a long-running-mode-only concern.
- **\"Double notifications for \`scheduling/accept\`\"** — \`emitEvent\` always passes \`envelope.message.method\` which is \`action/request\` or \`action/result\`, never \`scheduling/*\`. The classifier's \`scheduling/*\` cases are unreachable dead code; \`onMeetingConfirmed\` is the only source of a scheduling/accept notification. No duplicate.
- **\"Partial runtime state on failed identity start\"** — \`TapMessagingService.start()\` is idempotent (\`if (this.running) return\`), so retries through the cached \`ManagedTapRuntime\` work correctly. \`restartRuntime\` explicitly drops and recreates.
- **\"\`resolvePending\` type guard\"** — the method already routes via \`entry.metadata?.type\` and \`respondMeeting\` pre-filters through \`findPendingSchedulingEntry\`. A defensive assertion would add noise without catching a real scenario.
- **\"\`tap_gateway\` error wrapping leaks stack traces\"** — started implementing structured error envelopes, but \`packages/openclaw-plugin/test/tool.test.ts\` explicitly encodes the contract \"tool errors throw instead of flattening into success payloads.\" The OpenClaw SDK is responsible for formatting thrown errors; the audit's leak concern was not verified against the SDK. Reverted.

## Deferred (out of scope — flagged for follow-up)

- **\`create_invite\` in plugin violates thin-plugin rule** — per CLAUDE.md, invite generation is pure crypto and belongs in CLI only. Removing it is a user-facing breaking change and should be a separate PR with explicit agreement.
- **Silent auto-generation of identity \`name\`** — config UX change that could break existing Gateway deployments.
- **No TTL on pending request journal** — design decision (how long is \"too long\" for a pending request?).
- **Stale lock corruption recovery edge case** — needs fault-injection tests.
- **Heartbeat-wake fire-and-forget race** — needs broader reliability design.

## Test plan
- [x] \`bun run typecheck\` clean
- [x] \`bun run lint\` clean
- [x] \`bun run test\` — 683 passing, 38 skipped
- [ ] Smoke test inside a real OpenClaw Gateway with a configured TAP identity: send a scheduling proposal that requires approval and confirm the operator sees the \`[TAP Notifications]\` escalation.
- [ ] Smoke test \`tap_gateway transfer amount=abc\` — should fail fast with \"amount must be a positive number\" instead of a viem error.
- [ ] Smoke test \`TAP_OWS_WALLET=...\` env override inside Gateway is honored.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Medium risk because it changes runtime reconciliation/notification behavior and adds new drop/overflow handling that could affect operator visibility of pending approvals, though changes are localized and covered by tests.
> 
> **Overview**
> Fixes several reliability issues in the OpenClaw TAP plugin: deferred scheduling proposals now enqueue an *escalation* notification (mirroring transfer approvals) so operators actually see meeting requests that require approval.
> 
> Hardens background operation by adding an in-flight guard to periodic `syncOnce()` runs (avoids mutex pileups) and by bounding the `TapNotificationQueue` escalation overflow with a hard cap plus a synthetic `tap/queue-overflow` sentinel emitted on `drain()` when escalations were dropped.
> 
> Improves safety/observability by validating `amount` inputs for `transfer`/`request_funds` up front, logging failures to release the transport owner lock in core `TapMessagingService`, and documenting the `transfer` action; adds tests for the new queue overflow and amount validation behaviors.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit d2f3186dd86dc348e427642a0b232ff581f179ca. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by ggonzalez94** [comment]
## Codex adversarial review addressed (f052e48)

Re-verified each finding and pushed a follow-up commit.

### Fixed

- **OWS env overrides (high)** — `process.env.TAP_OWS_WALLET` / `TAP_OWS_API_KEY` plumbing in `registry.ts` reverted. In a multi-identity plugin host, those env vars are process-global and would force every identity onto the same wallet/API key. Per-dataDir YAML is the correct source inside the plugin; the env-override pattern remains valid in the CLI (single-identity). This also withdraws the original audit Finding #5 — it was wrong once you account for the single-process / multi-identity model.

- **Overflow dropped escalations hidden from agent (medium)** — `TapNotificationQueue` now tracks a `droppedEscalationCount` and emits a synthetic overflow sentinel at drain time: `"TAP notification queue dropped N escalations due to overflow. Run tap_gateway list_pending to recover missed approvals..."`. Stable messageId, counter resets on drain. Added two new tests: sentinel is emitted with correct count and oneLiner, no sentinel when no overflow.

### Deferred (predates this PR)

- **`tap_gateway transfer` is a no-confirmation on-chain spend path (critical)** — real concern, but the action predates this PR (`git log -S'"transfer"' -- packages/openclaw-plugin/src/tool.ts` → commit 4ad65c5 / PR #28). This branch only added input validation and the skill table row. Removing or gating the action is a design decision (remove entirely? add two-step approve? grant-based like transfer/request?) that warrants its own PR. Logged for follow-up — not changing it here.

### Test status

`bun run typecheck && bun run lint && bun run test` all green. **685 passing** / 38 skipped (two new notification-queue tests).

File: packages/openclaw-plugin/src/notification-queue.ts [hunk 1]
```diff
@@ -36,7 +50,26 @@ export class TapNotificationQueue {
...
+		if (this.droppedEscalationCount > 0) {
+			// Always emit a synthetic overflow notification so the agent is told
+			// to reconcile via list_pending. Without this, dropped escalations
+			// would leave pending approvals in the journal with no visible signal
+			// in the agent's [TAP Notifications] context.
+			const count = this.droppedEscalationCount;
+			this.droppedEscalationCount = 0;
+			drained.push({
+				type: "escalation",
+				identity: "",
```
[the remark this answers is not in the document]
Comment (ggonzalez94) [reply packages/openclaw-plugin/src/notification-queue.ts]: Good catch — fixed in df83f87. Added `identity` to `TapNotificationQueueOptions` and the registry now passes `identity: name` when constructing the queue. The sentinel now matches the rest of the drain batch. Test updated to assert the identity is propagated.
